proof/courier.go

```diff
@@ -849,6 +921,206 @@ func (h *HashMailCourier) SetSubscribers(
 // proof.Courier interface.
 var _ Courier = (*HashMailCourier)(nil)
 
+// UniverseRpcCourier is a universe RPC proof courier service handle. It
+// implements the Courier interface.
+type UniverseRpcCourier struct {
+	// recipient describes the recipient of the proof.
+	recipient Recipient
+
+	// client is the RPC client that the courier will use to interact with
+	// the universe RPC server.
+	client unirpc.UniverseClient
+
+	// deliveryLog is the log that the courier will use to record the
+	// attempted delivery of proofs to the receiver.
+	deliveryLog DeliveryLog
+
+	// subscribers is a map of components that want to be notified on new
+	// events, keyed by their subscription ID.
+	subscribers map[uint64]*fn.EventReceiver[fn.Event]
+
+	// subscriberMtx guards the subscribers map and access to the
+	// subscriptionID.
+	subscriberMtx sync.Mutex
+}
+
+// DeliverProof attempts to delivery a proof file to the receiver.
+func (c *UniverseRpcCourier) DeliverProof(ctx context.Context,
+	annotatedProof *AnnotatedProof) error {
+
+	// Decode annotated proof into proof file.
+	proofFile := &File{}
+	err := proofFile.Decode(bytes.NewReader(annotatedProof.Blob))
+	if err != nil {
+		return err
+	}
+
+	// Iterate over each proof in the proof file and submit to the courier
+	// service.
+	for i := 0; i < proofFile.NumProofs(); i++ {
+		transitionProof, err := proofFile.ProofAt(uint32(i))
+		if err != nil {
+			return err
+		}
+		proofAsset := transitionProof.Asset
+
+		// Construct asset leaf.
+		rpcAsset, err := taprpc.MarshalAsset(
+			ctx, &proofAsset, true, true, nil,
+		)
+		if err != nil {
+			return err
+		}
+
+		var proofBuf bytes.Buffer
+		if err := transitionProof.Encode(&proofBuf); err != nil {
+			return fmt.Errorf("error encoding proof file: %w", err)
+		}
+
+		assetLeaf := unirpc.AssetLeaf{
+			Asset:         rpcAsset,
+			IssuanceProof: proofBuf.Bytes(),
+		}
+
+		// Construct universe key.
+		outPoint := transitionProof.OutPoint()
+		assetKey := unirpc.MarshalAssetKey(
+			outPoint, proofAsset.ScriptKey.PubKey,
+		)
+		universeID := unirpc.MarshalUniverseID(proofAsset.ID())
+		universeKey := unirpc.UniverseKey{
+			Id:      universeID,
+			LeafKey: assetKey,
+		}
+
+		// Submit proof to courier.
+		_, err = c.client.InsertProof(ctx, &unirpc.AssetProof{
```

Comment:
We'll want to double check what the return is here if the proof already exists. To avoid bailing out early when not needed.

Reply:
The `testBasicSendUnidirectional` itest covers this because it performs two different send events for the same asset (single genesis), it spends the change. I've also manually tested that `InsertProof` can correctly handle a proof that it's seen before.